sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala

```diff
@@ -17,18 +17,270 @@
+ *           |      +-- 00008-[uuid3].log
+ *           |      +-- 00013-[uuid4].log
+ *           +-- 10.zip
+ *           |      +-- metadata         <--- contains mapping between 00007.sst and [uuid1].sst,
+ *                                            and the mapping between 00008.log and [uuid3].log
+ *           |      +-- OPTIONS-000005
+ *           |      +-- MANIFEST-000008
+ *           |      +-- CURRENT
+ *           |      ...
+ *           |
+ *           +-- 9.zip
+ *           +-- 8.zip
+ *           ...
+ *
+ * Note the following.
+ * - Each [version].zip is a complete description of all the data and metadata needed to recover
+ *   a RocksDB instance at the corresponding version. This is unlike the [version].delta files of
+ *   HDFSBackedStateStore where previous delta files needs to be read to be recovered.
+ * - This is safe wrt speculatively executed tasks running concurrently in different executors
+ *   as each task would upload a different copy of the generated immutable files and
+ *   atomically update the [version].zip.
+ * - Immutable files are identified uniquely based on their file name and file size.
+ * - Immutable files can be reused only across adjacent checkpoints/versions.
+ * - This class is thread-safe. Specifically, it is safe to concurrently delete old files from a
+ *   different thread than the task thread saving files.
+ *
+ * @param dfsRootDir  Directory where the [version].zip files will be stored
+ * @param localTempDir Local directory for temporary work
+ * @param hadoopConf   Hadoop configuration for talking to DFS
+ * @param loggingId    Id that will be prepended in logs for isolating concurrent RocksDBs
+ */
+class RocksDBFileManager(
+  dfsRootDir: String,
+  localTempDir: File,
+  hadoopConf: Configuration,
+  loggingId: String = "")
+  extends Logging {
+
+  import RocksDBImmutableFile._
+
+  private val versionToRocksDBFiles = new ConcurrentHashMap[Long, Seq[RocksDBImmutableFile]]
+  private lazy val fm = CheckpointFileManager.create(new Path(dfsRootDir), hadoopConf)
+  private val fs = new Path(dfsRootDir).getFileSystem(hadoopConf)
+
+  /** Save all the files in given local checkpoint directory as a committed version in DFS */
+  def saveCheckpointToDfs(checkpointDir: File, version: Long, numKeys: Long): Unit = {
+    logFilesInDir(checkpointDir, s"Saving checkpoint files for version $version")
+    val allCheckpointFiles = listRocksDBFiles(checkpointDir)
+    val (localImmutableFiles, localOtherFiles) = allCheckpointFiles.partition(isImmutableFile)
+    val rocksDBFiles = saveImmutableFilesToDfs(version, checkpointDir, localImmutableFiles)
+    val metadata = RocksDBCheckpointMetadata(rocksDBFiles, numKeys)
+    rememberImmutableFiles(version, rocksDBFiles)
+    val metadataFile = localMetadataFile(checkpointDir)
+    metadata.writeToFile(metadataFile)
+    logInfo(s"Written metadata for version $version:\n${metadata.prettyJson}")
+
+    if (version <= 0 && numKeys == 0) {
```

Comment:
We also use batch ID + 1 as the version for the RocksDB state store. Since the batch ID starts from -1 so for the corner case of the first empty batch (batch ID 0), we add the safeguard here to make sure the working dir is created.

Reply:
What about simply making sure the directory has been created in prior? We can add below in initialization and don't bother with handling specific case.

```
  // create a root directory on initialization if necessary
  if (!fm.exists(new Path(dfsRootDir))) {
    fm.mkdirs(new Path(dfsRootDir))
  }
```

If we'd like to avoid paying the cost on calling exist, let's keep the code and change the version to 1 as I met the case the version starts from 1 during test.

Reply:
Yeah, let's change the version to 1 for avoiding the extra cost of file operations. Will change it in the next commit.